### ISSUE TITLE
--regex='\0' fails for passwords with non-ascii characters

### ISSUE THREAD

**Issue by frank-dittrich** [magnumripper/JohnTheRipper#1976]
latest bleeding-jumbo, latest test suite version, latest rexgen master commit with this patch to drop librexgen debug output:
```
diff --git a/src/librexgen/string/simplestring.cpp b/src/librexgen/string/simplestring.cpp
index 9c52188..0e1d5de 100644
--- a/src/librexgen/string/simplestring.cpp
+++ b/src/librexgen/string/simplestring.cpp
@@ -45,7 +45,7 @@ void SimpleString::toupper(unsigned int n) {
 }
 
 void SimpleString::append(const char* ch, size_t ch_len) {
-  fprintf(stderr, "APPEND\n");
+//  fprintf(stderr, "APPEND\n");
        /* ensure that buffer is not full */
        if (length > SIMPLESTRING_MAXLEN) {
                return;
@@ -60,7 +60,7 @@ void SimpleString::append(const char* ch, size_t ch_len) {
        for (size_t idx=0; idx<ch_len; ++idx) {
                characters[length++] = ch[idx];
 
-    fprintf(stderr, "%04x\n", characters[length-1].codepoint);
+//    fprintf(stderr, "%04x\n", characters[length-1].codepoint);
        }
 }
```

```
(master)test $ ./jtrts.pl -q -passthru="--regex='\0'"
-------------------------------------------------------------------------------
- JtR-TestSuite (jtrts). Version 1.13, Dec 21, 2014.  By, Jim Fougeron & others
- Testing:  John the Ripper 1.8.0.6-jumbo-1-1814-gca11872 [linux-gnu 64-bit AVX-ac]
--------------------------------------------------------------------------------
form=dynamic_0                    guesses: 1195 -show=1197 0:00:00:00 DONE : Expected count(s) (1500)(-show1502)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_0                    guesses:  876 -show= 876 0:00:00:00 DONE : Expected count(s) (1500)(920)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_0                    guesses: 1436 -show=1436 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_0-raw                guesses: 1195 -show=1195 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_1                    guesses: 1200 -show=1200 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_1                    guesses:  887 -show= 887 0:00:00:00 DONE : Expected count(s) (1500)(923)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_2                    guesses: 1195 -show=1195 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_2                    guesses:  876 -show= 876 0:00:00:00 DONE : Expected count(s) (1500)(920)  [!!!FAILED1!!!  ret_val=0]
form=dynamic_2-raw                guesses: 1195 -show=1195 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
...
form=mssql05-koi8r-gen            guesses:  385 -show= 385 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=MSSql05-koi8r                guesses:    0 -show=   0 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=oracle-koi8r                 guesses:    0 -show=   0 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=SAP F/G                      guesses: 1786 -show=1786 0:00:00:00 DONE : Expected count(s) (6794)  [!!!FAILED1!!!  ret_val=0]
form=sapB                         guesses: 1404 -show=1404 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=gost                         guesses: 1195 -show=1195 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=gost                         guesses: 1195 -show=1195 0:00:00:00 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=tc_aes_xts (3-types)         guesses:  968 -show= 968 0:00:00:21 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
form=tc_sha512                    guesses:  960 -show= 960 0:00:00:15 DONE : Expected count(s) (1500)  [!!!FAILED1!!!  ret_val=0]
Some tests had Errors. Performed 599 tests.  589 errors  1 errors reprocessing the .POT files
Time used was 1400 seconds
```

John is always missing the passwords with single-byte non-ascii characters.

If there's a way to make john crack such passwords with --regex, then it should be documented, and the test suite should possibly be enhanced to work with `-passthru="--regex='\0'"`. 

But it almost looks like we need to convert everything into utf-8 before passing it to librexgen, and then convert it back into the desired encoding.
But that wouldn't work with various single-byte encodings mixed into the same input word list.

**Comment by frank-dittrich** [comment]
Not sure if there's a way to pass the input encoding to librexgen.

rexgen seems to allow just 5 encodings (see `rexgen -h` output):
```
   -u8:       encode values in UTF-8
   -u16[le]:  encode values in UTF-16BE (resp. UTF-16LE)
   -u32[le]:  encode values in UTF-32BE (resp. UTF-32LE)
```

May be this can be addressed in librexgen? If there are non-ascii bytes which are not part of a valid utf-8 character, just leave those bytes unchanged?

**Comment by frank-dittrich** [comment]
This
```
form=SAP F/G                      guesses: 1786 -show=1786 0:00:00:00 DONE : Expected count(s) (6794)  [!!!FAILED1!!!  ret_val=0]
```
seems to be a separate problem (definitely a librexgen issue).

Looks like librexgen is not handling utf-8 characters which are encoded using more than 2 bytes.
E.g., it missed `€ä34567890123456789012345678901234567890`, I guess due to the € character.

**Comment by frank-dittrich** [comment]
Looks like not even `rexgen -u8` works with non-ascii UTF-8 characters, see https://github.com/teeshop/rexgen/issues/26

**Comment by magnumripper** [comment]
> May be this can be addressed in librexgen? If there are non-ascii bytes which are not part of a valid utf-8 character, just leave those bytes unchanged?

It's easier to say so than to do it: I presume rexgen uses UTF-32 internally (which is a good idea). You could of course convert eg. a stray 0xEE to U+00EE, that's easy. But it will definitely be converted back to proper UTF-8 from **that** in the end - which is definitely something very different than the original.

**Comment by jfoug** [comment]
There really should be a -raw  (or -u0) added to the endcoding for the lib, but that may not be all that easy to do.

**Comment by frank-dittrich** [comment]
You can't easily "toggle case" for non-ascii characters with -raw.

**Comment by magnumripper** [comment]
You can't, period. So we're not even trying. In Jumbo raw is just ASCII plus non-destructive handling of others. For Unicode formats, raw assumes ISO-8859-1 (ie. dumb cast to 16-bit).

**Comment by magnumripper** [comment]
In my opinion, Jan should not bother making silly workarounds for non-UTF-8. Crap in, crap out. He should just fix the 3+ byte UTF-8 bug.

**Comment by frank-dittrich** [comment]
Looks like we have to wait for https://github.com/teeshop/rexgen/issues/26 being addressed before we can close this issue.

**Comment by magnumripper** [comment]
I think this is mostly fixed now in rexgen.
```
$ rexgen m[üö]ller
müller
möller

$ ../run/john -stdout -regex=m[üö]ller
müller
möller
2p 0:00:00:00 0.00% 200.0p/s möller
```
However, that TS test still fails (although with fewer failures). This isn't something we'll ever be able to fix since our "--enc=raw" simply isn't compatible with librexgen's proper handling of Unicode. Basically the problem is [what I said above](https://github.com/magnumripper/JohnTheRipper/issues/1976#issuecomment-168540846).

So I'm closing this. If you find problems with "proper input", that is either using UTF-8 input or some codepage (and using our codepage options accordingly), please open a new issue.

**Comment by jfoug** [comment]
You have to remember.  The wordlist in TS is PURPOSELY crap.  I took it and splattered in simply setting high bit to whatever character was there  all over the place.  This did flush out a lot of issues along the way in john.

However, even in things like pass_gen, we had to FORCE perl to not try to do what it wanted to do all the time, and FORCE it to simply treat the data as a stream of bytes.   rexgen 'used' to work that way, but Jan made it always assume that the data is 'good', and of a single encoding, so now it will not properly handle arbitrary 8 bit data (in a raw mode).   

rexgen is what it is.
